File: v4/rust/iprange-livedb/src/draft_store/feed_merge.rs
//! One-sweep exact named-feed replacement over normalized mapped ranges.

use crate::bootstrap::Bootstrap;
use crate::cancellation::CancellationToken;
use crate::cardinality::Cardinality129;
use crate::contract::{AddressFamily, MembershipOperation, ValueKind};
use crate::error::{Error, Result};
use crate::key::{IpKey, Ipv4Key, Ipv6Key};
use crate::membership_dictionary::Interned;
use crate::range_bulk::{Builder, Record};
use crate::range_cursor::DirectRange;
use crate::range_mutation;
use crate::range_store_cursor::Cursor;
use crate::workflow::{compare::ScannedComparison, Comparison};

use super::DraftStore;

pub(crate) struct FeedMerge {
    pub(crate) input_intervals: u64,
    pub(crate) input_addresses: Cardinality129,
    pub(crate) comparison: ScannedComparison,
}

impl DraftStore<'_> {
    pub(crate) fn add_feed_coverage<K: IpKey>(&mut self, from: K, to: K) -> Result<()> {
        let mut root = self.draft.workflow_range_root;
        let mut count = self.draft.workflow_range_count;
        range_mutation::assign_private_untracked(self, &mut root, &mut count, from, to, 1)?;
        self.draft.workflow_range_root = root;
        self.draft.workflow_range_count = count;
        Ok(())
    }

    pub(crate) fn merge_feed(
        &mut self,
        base: &Bootstrap,
        member: Interned,
        create: bool,
        cancellation: &CancellationToken,
    ) -> Result<FeedMerge> {
        if create && self.draft.workflow_range_root == 0 {
            return Ok(FeedMerge {
                input_intervals: 0,
                input_addresses: Cardinality129::ZERO,
                comparison: ScannedComparison {
                    comparison: Comparison::default(),
                    before_intervals: 0,
                    after_intervals: 0,
                },
            });
        }
        match base.meta.address_family {
            AddressFamily::Ipv4 => self.merge_feed_family::<Ipv4Key>(base, member, cancellation),
            AddressFamily::Ipv6 => self.merge_feed_family::<Ipv6Key>(base, member, cancellation),
        }
    }

    fn merge_feed_family<K: IpKey>(
        &mut self,
        base: &Bootstrap,
        member: Interned,
        cancellation: &CancellationToken,
    ) -> Result<FeedMerge> {
        cancellation.check()?;
        let mut coverage_meta = self.draft.meta;
        coverage_meta.range_root = self.draft.workflow_range_root;
        coverage_meta.range_record_count = self.draft.workflow_range_count;
        let old = Cursor::<K>::new(self, &base.meta, false)?;
        let coverage = Cursor::<K>::new(self, &coverage_meta, true)?;
        let mut merge = Merge::new(old, coverage, self.draft.meta.txn_id, member);
        let (root, count, result) = merge.run(self, cancellation)?;
        range_mutation::retire_tree::<K, _, _>(self, base.meta.range_root, || {
            cancellation.check()
        })?;
        self.draft.meta.range_root = root;
        self.draft.meta.range_record_count = count;
        self.draft.workflow_range_root = 0;
        self.draft.workflow_range_count = 0;
        Ok(result)
    }
}

struct Merge<K> {
    old_cursor: Cursor<K>,
    coverage_cursor: Cursor<K>,
    old: Option<DirectRange<K>>,
    coverage: Option<DirectRange<K>>,
    output: Output<K>,
    member: Interned,
    cached: Option<CachedMembership>,
    input_intervals: u64,
    input_addresses: Cardinality129,
    projection: Projection<K>,
}

#[derive(Clone, Copy)]
struct CachedMembership {
    old: u32,
    covered: bool,
    new: u32,
}

impl<K: IpKey> Merge<K> {
    fn new(old: Cursor<K>, coverage: Cursor<K>, transaction: u64, member: Interned) -> Self {
        Self {
            old_cursor: old,
            coverage_cursor: coverage,
            old: None,
            coverage: None,
            output: Output::new(transaction),
            member,
            cached: None,
            input_intervals: 0,
            input_addresses: Cardinality129::ZERO,
            projection: Projection::default(),
        }
    }

    fn run(
        &mut self,
        store: &mut DraftStore<'_>,
        cancellation: &CancellationToken,
    ) -> Result<(u32, u64, FeedMerge)> {
        self.advance_old(store, cancellation)?;
        self.advance_coverage(store, cancellation)?;
        while self.old.is_some() || self.coverage.is_some() {
            cancellation.check()?;
            self.step(store, cancellation)?;
        }
        let (root, count) = self.output.finish(store)?;
        let comparison = std::mem::take(&mut self.projection).finish(self.input_addresses)?;
        Ok((
            root,
            count,
            FeedMerge {
                input_intervals: self.input_intervals,
                input_addresses: self.input_addresses,
                comparison,
            },
        ))
    }

    fn step(&mut self, store: &mut DraftStore<'_>, cancellation: &CancellationToken) -> Result<()> {
        let Some(old) = self.old else {
            let coverage = self
                .coverage
                .ok_or(Error::Corrupt("feed merge lost both inputs"))?;
            self.emit(store, coverage.from, coverage.to, None, true)?;
            return self.advance_coverage(store, cancellation);
        };
        let Some(coverage) = self.coverage else {
            self.emit(store, old.from, old.to, Some(old.value), false)?;
            return self.advance_old(store, cancellation);
        };
        if old.to < coverage.from {
            self.emit(store, old.from, old.to, Some(old.value), false)?;
            return self.advance_old(store, cancellation);
        }
        if coverage.to < old.from {
            self.emit(store, coverage.from, coverage.to, None, true)?;
            return self.advance_coverage(store, cancellation);
        }
        self.overlap(store, old, coverage, cancellation)
    }

    fn overlap(
        &mut self,
        store: &mut DraftStore<'_>,
        mut old: DirectRange<K>,
        mut coverage: DirectRange<K>,
        cancellation: &CancellationToken,
    ) -> Result<()> {
        if old.from < coverage.from {
            let end = previous(coverage.from, "feed old prefix")?;
            self.emit(store, old.from, end, Some(old.value), false)?;
            old.from = coverage.from;
            self.old = Some(old);
            return Ok(());
        }
        if coverage.from < old.from {
            let end = previous(old.from, "feed coverage prefix")?;
            self.emit(store, coverage.from, end, None, true)?;
            coverage.from = old.from;
            self.coverage = Some(coverage);
            return Ok(());
        }

        let end = old.to.min(coverage.to);
        self.emit(store, old.from, end, Some(old.value), true)?;
        if old.to == end {
            self.advance_old(store, cancellation)?;
        } else {
            old.from = next(end, "feed old remainder")?;
            self.old = Some(old);
        }
        if coverage.to == end {
            self.advance_coverage(store, cancellation)
        } else {
            coverage.from = next(end, "feed coverage remainder")?;
            self.coverage = Some(coverage);
            Ok(())
        }
    }

    fn emit(
        &mut self,
        store: &mut DraftStore<'_>,
        from: K,
        to: K,
        old: Option<u32>,
        covered: bool,
    ) -> Result<()> {
        let (new, before) = self.transform(store, old, covered)?;
        self.projection.observe(from, to, before, covered)?;
        if new != 0 {
            self.output.emit(
                store,
                Record {
                    from,
                    to,
                    value: new,
                },
            )?;
        }
        Ok(())
    }

    fn transform(
        &mut self,
        store: &mut DraftStore<'_>,
        old: Option<u32>,
        covered: bool,
    ) -> Result<(u32, bool)> {
        let Some(old) = old else {
            return Ok((if covered { self.member.id } else { 0 }, false));
        };
        let new = if let Some(cached) = self
            .cached
            .filter(|cached| cached.old == old && cached.covered == covered)
        {
            cached.new
        } else {
            let operation = if covered {
                MembershipOperation::Union
            } else {
                MembershipOperation::Difference
            };
            let new = store
                .combine_memberships(old, self.member.id, self.member.word_count, operation)?
                .unwrap_or(0);
            self.cached = Some(CachedMembership { old, covered, new });
            new
        };
        let before = if covered { new == old } else { new != old };
        Ok((new, before))
    }

    fn advance_old(
        &mut self,
        store: &mut DraftStore<'_>,
        cancellation: &CancellationToken,
    ) -> Result<()> {
        cancellation.check()?;
        self.old = self.old_cursor.next(store)?;
        if let Some(range) = self.old {
            store.track_membership_refcount(range.value, -1)?;
        }
        Ok(())
    }

    fn advance_coverage(
        &mut self,
        store: &mut DraftStore<'_>,
        cancellation: &CancellationToken,
    ) -> Result<()> {
        cancellation.check()?;
        self.coverage = self.coverage_cursor.next(store)?;
        if let Some(range) = self.coverage {
            self.input_intervals = self
                .input_intervals
                .checked_add(1)
                .ok_or(Error::ArithmeticOverflow("feed input intervals"))?;
            self.input_addresses = add(self.input_addresses, length(range.from, range.to)?)?;
        }
        Ok(())
    }
}

struct Output<K> {
    builder: Builder<K>,
    pending: Option<Record<K>>,
}

impl<K: IpKey> Output<K> {
    fn new(transaction: u64) -> Self {
        Self {
            builder: Builder::new(transaction, ValueKind::Membership),
            pending: None,
        }
    }

    fn emit(&mut self, store: &mut DraftStore<'_>, record: Record<K>) -> Result<()> {
        if let Some(pending) = self.pending.as_mut() {
            if pending.value == record.value && pending.to.checked_next() == Some(record.from) {
                pending.to = record.to;
                crate::work::range_coalesced(1);
                return Ok(());
            }
        }
        if let Some(pending) = self.pending.replace(record) {
            self.push(store, pending)?;
        }
        Ok(())
    }

    fn finish(&mut self, store: &mut DraftStore<'_>) -> Result<(u32, u64)> {
        if let Some(record) = self.pending.take() {
            self.push(store, record)?;
        }
        self.builder.finish(store)
    }

    fn push(&mut self, store: &mut DraftStore<'_>, record: Record<K>) -> Result<()> {
        store.track_membership_refcount(record.value, 1)?;
        self.builder.push(store, record)
    }
}

struct Projection<K> {
    result: Comparison,
    before_intervals: u64,
    after_intervals: u64,
    last_to: Option<K>,
    last_before: bool,
    last_after: bool,
}

impl<K> Default for Projection<K> {
    fn default() -> Self {
        Self {
            result: Comparison::default(),
            before_intervals: 0,
            after_intervals: 0,
            last_to: None,
            last_before: false,
            last_after: false,
        }
    }
}

impl<K: IpKey> Projection<K> {
    fn observe(&mut self, from: K, to: K, before: bool, after: bool) -> Result<()> {
        let adjacent = self.last_to.and_then(IpKey::checked_next) == Some(from);
        let count = length(from, to)?;
        if before {
            self.result.before = add(self.result.before, count)?;
            if !adjacent || !self.last_before {
                self.before_intervals = increment(self.before_intervals, "before feed intervals")?;
            }
        }
        if after {
            self.result.after = add(self.result.after, count)?;
            if !adjacent || !self.last_after {
                self.after_intervals = increment(self.after_intervals, "after feed intervals")?;
            }
        }
        match (before, after) {
            (true, true) => self.result.unchanged = add(self.result.unchanged, count)?,
            (true, false) => self.result.removed = add(self.result.removed, count)?,
            (false, true) => self.result.added = add(self.result.added, count)?,
            (false, false) => {}
        }
        self.last_to = Some(to);
        self.last_before = before;
        self.last_after = after;
        Ok(())
    }

    fn finish(self, input_addresses: Cardinality129) -> Result<ScannedComparison> {
        if self.result.after != input_addresses
            || self.after_intervals == 0 && input_addresses != Cardinality129::ZERO
        {
            return Err(Error::Corrupt(
                "feed merge input and output coverage disagree",
            ));
        }
        let before = add(self.result.unchanged, self.result.removed)?;
        let after = add(self.result.unchanged, self.result.added)?;
        if before != self.result.before
            || after != self.result.after
            || self.result.changed != Cardinality129::ZERO
        {
            return Err(Error::Corrupt("feed merge counters do not balance"));
        }
        Ok(ScannedComparison {
            comparison: self.result,
            before_intervals: self.before_intervals,
            after_intervals: self.after_intervals,
        })
    }
}

fn length<K: IpKey>(from: K, to: K) -> Result<Cardinality129> {
    from.inclusive_cardinality(to)
}

fn add(left: Cardinality129, right: Cardinality129) -> Result<Cardinality129> {
    left.checked_add(right)
        .map_err(|_| Error::ArithmeticOverflow("feed address count"))
}

fn increment(value: u64, context: &'static str) -> Result<u64> {
    value
        .checked_add(1)
        .ok_or(Error::ArithmeticOverflow(context))
}

fn previous<K: IpKey>(key: K, context: &'static str) -> Result<K> {
    key.checked_previous()
        .ok_or(Error::ArithmeticOverflow(context))
}

fn next<K: IpKey>(key: K, context: &'static str) -> Result<K> {
    key.checked_next().ok_or(Error::ArithmeticOverflow(context))
}

File: v4/go/internal/writer/feed_merge.go
// Exact named-feed policy over the authoritative ordered range merge
// (Rust draft_store/feed_merge.rs). The empty-map feed trio builds the
// workflow's constant-value ranges, the private coverage tree is the
// untracked input of one merge over the committed base, and the
// feed policy unions or differences the member bitmap with the base
// membership, projecting exact before/after address and interval counts.

package writer

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/tree"
)

// feedMerge is the per-feed outcome of one coverage merge (Rust
// FeedMerge).
type feedMerge struct {
	inputIntervals uint64
	inputAddresses format.Cardinality129
	comparison     scannedComparison
}

// beginEmptyMapFeed starts the empty-map workflow of a value-free base:
// the base and draft range trees must both be empty, and the range tree
// becomes draft-private (Rust DraftStore::begin_empty_map_feed).
func (s *DraftStore) beginEmptyMapFeed() error {
	if s.draft.base.RangeRoot != 0 || s.draft.base.RangeRecordCount != 0 ||
		s.draft.meta.RangeRoot != 0 || s.draft.meta.RangeRecordCount != 0 {
		return corrupt("empty-map feed range tree is not empty")
	}
	s.draft.rangeTreePrivate = true
	return nil
}

// addEmptyMapFeedRange pushes one constant member-valued range into the
// private draft tree (Rust DraftStore::add_empty_map_feed_range).
func (s *DraftStore) addEmptyMapFeedRange(from, to tree.Key, member membershipHandle, input *unionInput) error {
	memberID, _ := member.stored()
	return s.addPrivateConstantRange(from, to, memberID, input)
}

// finishEmptyMapFeedRanges seals the constant ranges, accounts the
// member refcount for every record of the sealed tree, and returns the
// ordered-prefix address count when one was built (Rust
// DraftStore::finish_empty_map_feed_ranges).
func (s *DraftStore) finishEmptyMapFeedRanges(member membershipHandle, input *unionInput) (format.Cardinality129, bool, error) {
	if err := s.finishPrivateConstantRanges(input); err != nil {
		return format.Cardinality129{}, false, err
	}
	ordered, hasOrdered := input.orderedAddresses()
	value, _ := member.stored()
	count := int64(s.draft.meta.RangeRecordCount)
	if count < 0 {
		return format.Cardinality129{}, false, overflow("membership range refcount")
	}
	if count != 0 {
		if err := s.trackMembershipRefcount(value, count); err != nil {
			return format.Cardinality129{}, false, err
		}
	}
	return ordered, hasOrdered, nil
}

// addFeedCoverage pushes one value-1 range into the private workflow
// coverage tree, untracked (Rust DraftStore::add_feed_coverage: the
// coverage tree is the input of the upcoming merge, never part of the
// published range tree).
func (s *DraftStore) addFeedCoverage(from, to tree.Key, input *unionInput) error {
	family, err := s.rangeFamily()
	if err != nil {
		return err
	}
	root := s.draft.workflowRangeRoot
	count := s.draft.workflowRangeCount
	ctx := &rangeCtx{family: family, store: s, root: &root, count: &count}
	if _, err := pushPrivateUntracked(ctx, from, to, 1, input); err != nil {
		return err
	}
	s.draft.workflowRangeRoot = root
	s.draft.workflowRangeCount = count
	return nil
}

// finishFeedCoverage seals the pending coverage input (Rust
// DraftStore::finish_feed_coverage).
func (s *DraftStore) finishFeedCoverage(input *unionInput) error {
	family, err := s.rangeFamily()
	if err != nil {
		return err
	}
	root := s.draft.workflowRangeRoot
	count := s.draft.workflowRangeCount
	ctx := &rangeCtx{family: family, store: s, root: &root, count: &count}
	if _, err := finishInputUntracked(ctx, input); err != nil {
		return err
	}
	s.draft.workflowRangeRoot = root
	s.draft.workflowRangeCount = count
	return nil
}

// mergeFeed merges the workflow coverage tree over the committed base
// generation, applying the member bitmap through the feed policy (Rust
// DraftStore::merge_feed). A created feed with an empty coverage tree
// returns the exact no-change result without scanning anything.
func (s *DraftStore) mergeFeed(base format.Meta, member membershipHandle, create bool, check func() error) (feedMerge, error) {
	if create && s.draft.workflowRangeRoot == 0 {
		return emptyFeedMerge(), nil
	}
	return s.mergeFeedFamily(base, member, check)
}

// mergeFeedFamily runs the coverage merge for the base address family
// (Rust DraftStore::merge_feed_family): the coverage meta is the draft
// meta with the workflow range roots substituted, and the policy output
// is the scanned before/after comparison.
func (s *DraftStore) mergeFeedFamily(base format.Meta, member membershipHandle, check func() error) (feedMerge, error) {
	coverageMeta := s.draft.meta
	coverageMeta.RangeRoot = s.draft.workflowRangeRoot
	coverageMeta.RangeRecordCount = s.draft.workflowRangeCount
	var codec rangeFamily
	if base.AddressFamily == format.AddressFamilyIPv4 {
		codec = rangeCodec4{}
	} else {
		codec = rangeCodec6{}
	}
	policy := newFeedPolicy(member, base.AddressFamily)
	inputIntervals, finished, err := mergeCoverage[scannedComparison](s, coverageMeta, base, codec, &policy, check, "feed input intervals")
	if err != nil {
		return feedMerge{}, err
	}
	s.draft.workflowRangeRoot = 0
	s.draft.workflowRangeCount = 0
	return feedMerge{
		inputIntervals: inputIntervals,
		inputAddresses: finished.comparison.after,
		comparison:     finished,
	}, nil
}

// emptyFeedMerge is the exact no-change outcome of an empty workflow
// (Rust empty_result).
func emptyFeedMerge() feedMerge {
	return feedMerge{comparison: scannedComparison{comparison: comparison{}}}
}

// cachedMembership is one cached transform outcome keyed on the old
// membership and the coverage presence (Rust CachedMembership). The new
// bitmap is the Option<u32> combine outcome, carried as optionalValue.
type cachedMembership struct {
	old     uint32
	covered bool
	new     optionalValue
}

// feedPolicy is the per-feed merge policy (Rust FeedPolicy): the member
// bitmap is unioned into covered segments and differenced out of
// uncovered segments, with one cache slot and the running projection.
type feedPolicy struct {
	member     membershipHandle
	cached     cachedMembership
	hasCached  bool
	projection feedProjection
	family     uint8
}

// newFeedPolicy starts one feed policy over the member handle and the
// merge address family (Rust FeedPolicy::new; the projection fixes its
// family at plan time like HistoryPolicy<K>).
func newFeedPolicy(member membershipHandle, family uint8) feedPolicy {
	return feedPolicy{member: member, family: family, projection: feedProjection{family: family}}
}

// transform returns the merged membership of one segment (Rust
// FeedPolicy::transform): no old bitmap adopts the member when covered
// and stays absent otherwise; an old bitmap combines with the member
// through the cached union/difference.
func (p *feedPolicy) transform(store *DraftStore, old, incoming optionalValue) (optionalValue, error) {
	covered := incoming.present
	memberID, memberWords := p.member.stored()
	if !old.present {
		if covered {
			return someValue(memberID), nil
		}
		return noneValue(), nil
	}
	if p.hasCached && p.cached.old == old.value && p.cached.covered == covered {
		return p.cached.new, nil
	}
	operation := membershipUnion
	if !covered {
		operation = membershipDifference
	}
	new, present, err := store.combineMemberships(old.value, memberID, memberWords, operation)
	if err != nil {
		return optionalValue{}, err
	}
	p.hasCached = true
	p.cached = cachedMembership{old: old.value, covered: covered, new: optionalValue{value: new, present: present}}
	return p.cached.new, nil
}

// observe projects one transformed segment into the before/after counts
// (Rust FeedPolicy::observe: the before classification compares the new
// bitmap with the old one around the coverage).
func (p *feedPolicy) observe(from, to tree.Key, old, incoming, new optionalValue) error {
	after := incoming.present
	var before bool
	if after {
		before = sameOptional(new, old)
	} else {
		before = !sameOptional(new, old)
	}
	return p.projection.observe(from, to, before, after)
}

// finish balances and returns the projection (Rust FeedPolicy::finish
// over Projection::finish).
func (p *feedPolicy) finish() (scannedComparison, error) {
	return p.projection.finish()
}

// preserveWithoutInput is false like every Rust policy except the
// explicit preserving ones; the feed policies always consume their
// coverage input (Rust Policy::PRESERVE_WITHOUT_INPUT default).
func (p *feedPolicy) preserveWithoutInput() bool { return false }

// feedProjection is the scanned before/after projection of one feed
// policy (Rust Projection<K>): the exact comparison counts plus the
// interval counters, with the adjacency state embedded by value so one
// observe never allocates.
type feedProjection struct {
	result          comparison
	beforeIntervals uint64
	afterIntervals  uint64
	lastTo          tree.Key
	hasLastTo       bool
	lastBefore      bool
	lastAfter       bool
	family          uint8
}

// observe folds one transformed segment into the projection (Rust
// Projection::observe): adjacent same-class segments share one interval
// count, and every address count is exact.
func (p *feedProjection) observe(from, to tree.Key, before, after bool) error {
	adjacent := false
	if p.hasLastTo {
		next, ok := nextKey(p.family, p.lastTo)
		adjacent = ok && next.Equal(from)
	}
	count, err := familyInclusiveCardinality(p.family, from, to)
	if err != nil {
		return err
	}
	if before {
		p.result.before, err = p.result.before.Add(count)
		if err != nil {
			return overflow("ordered merge address count")
		}
		if !adjacent || !p.lastBefore {
			p.beforeIntervals, err = incrementFeedIntervals(p.beforeIntervals, "before feed intervals")
			if err != nil {
				return err
			}
		}
	}
	if after {
		p.result.after, err = p.result.after.Add(count)
		if err != nil {
			return overflow("ordered merge address count")
		}
		if !adjacent || !p.lastAfter {
			p.afterIntervals, err = incrementFeedIntervals(p.afterIntervals, "after feed intervals")
			if err != nil {
				return err
			}
		}
	}
	switch {
	case before && after:
		p.result.unchanged, err = p.result.unchanged.Add(count)
		if err != nil {
			return overflow("ordered merge address count")
		}
	case before && !after:
		p.result.removed, err = p.result.removed.Add(count)
		if err != nil {
			return overflow("ordered merge address count")
		}
	case !before && after:
		p.result.added, err = p.result.added.Add(count)
		if err != nil {
			return overflow("ordered merge address count")
		}
	}
	p.lastTo = to
	p.hasLastTo = true
	p.lastBefore = before
	p.lastAfter = after
	return nil
}

// finish balances the six counts and returns the scanned comparison
// (Rust Projection::finish): the before/after totals must equal the
// unchanged/removed and unchanged/added sums, the changed class must be
// empty, and the interval counter must agree with the address count
// emptiness.
func (p *feedProjection) finish() (scannedComparison, error) {
	unchangedRemoved, err := p.result.unchanged.Add(p.result.removed)
	if err != nil {
		return scannedComparison{}, overflow("ordered merge address count")
	}
	unchangedAdded, err := p.result.unchanged.Add(p.result.added)
	if err != nil {
		return scannedComparison{}, overflow("ordered merge address count")
	}
	if unchangedRemoved.Compare(p.result.before) != 0 ||
		unchangedAdded.Compare(p.result.after) != 0 ||
		p.result.changed.Compare(format.CardinalityZero()) != 0 ||
		(p.afterIntervals == 0) != (p.result.after.Compare(format.CardinalityZero()) == 0) {
		return scannedComparison{}, corrupt("feed merge counters do not balance")
	}
	return scannedComparison{
		comparison:      p.result,
		beforeIntervals: p.beforeIntervals,
		afterIntervals:  p.afterIntervals,
	}, nil
}

// incrementFeedIntervals adds one interval count with the Rust overflow
// class (Rust increment).
func incrementFeedIntervals(value uint64, context string) (uint64, error) {
	return checkedAdd(value, 1, context)
}
